### PR TITLE
[BEAM-2899] Implement most of JobApi#run

### PR DESCRIPTION
**Pull request by tgroh**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by tgroh** [comment]
CC: @herohde

File: runners/reference/java/pom.xml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    <groupId>org.apache.beam</groupId>
+    <artifactId>beam-runners-reference-parent</artifactId>
+    <version>2.3.0-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
+  </parent>
+
+  <artifactId>beam-runners-reference-java</artifactId>
+
+  <name>Apache Beam :: Runners :: Reference :: Java</name>
+
```
Comment (lukecwik) [inline runners/reference/java/pom.xml]: add a description

File: runners/reference/java/pom.xml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>beam-runners-core-construction-java</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>com.google.protobuf</groupId>
+      <artifactId>protobuf-java</artifactId>
```
Comment (lukecwik) [inline runners/reference/java/pom.xml]: Dependencies between build.gradle and pom.xml here don't match on protobuf-java/jsr305/hamcrest core/junit/slf4j-jdk14/slf4j-api. Please prune unneeded dependencies and also do we need to use provided dependencies?

**Comment by tgroh** [comment]
Updated dependencies on both ends.

**Comment by lukecwik** [comment]
Did you forget to push an update, I still see hamcrest in runners/reference/java/pom.xml but not in runners/reference/java/build.gradle?

**Comment by tgroh** [comment]
Somehow I missed something.

**Comment by lukecwik** [comment]
LGTM, self merge when able.

**Comment by lukecwik** [comment]
Gradle build is failing on the file you added: 
Applying build_rules.gradle to beam-runners-reference-java
applyJavaNature with default configuration for project beam-runners-reference-java

FAILURE: Build failed with an exception.

* Where:
Build file '/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_GradleBuild/src/runners/reference/java/build.gradle' line: 30

* What went wrong:
A problem occurred evaluating project ':beam-runners-parent:beam-runners-reference-parent:beam-runners-reference-java'.
> Cannot convert a null value to an object of type Dependency.
  The following types/formats are supported:
    - Instances of Dependency.
    - String or CharSequence values, for example 'org.gradle:gradle-core:1.0'.
    - Maps, for example [group: 'org.gradle', name: 'gradle-core', version: '1.0'].
    - FileCollections, for example files('some.jar', 'someOther.jar').
    - Projects, for example project(':some:project:path').
    - ClassPathNotation, for example gradleApi().

File: runners/reference/java/build.gradle [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+description = "Apache Beam :: Runners :: Reference :: Java"
+
+dependencies {
+  shadow project(path: ":beam-model-parent:beam-model-pipeline", configuration: "shadow")
+  shadow project(path: ":beam-runners-parent:beam-runners-core-construction-java", configuration: "shadow")
+  shadow library.java.slf4j_api
+  testCompile library.java.junit
+  testCompile library.java.slf4j_jdk14
+  testCompile library.java.hamcrest
```
Comment (lukecwik) [inline runners/reference/java/build.gradle]: library.java.hamcrest_core?

Applying build_rules.gradle to beam-runners-reference-java
applyJavaNature with default configuration for project beam-runners-reference-java

FAILURE: Build failed with an exception.

* Where:
Build file '/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_GradleBuild/src/runners/reference/java/build.gradle' line: 30

* What went wrong:
A problem occurred evaluating project ':beam-runners-parent:beam-runners-reference-parent:beam-runners-reference-java'.
> Cannot convert a null value to an object of type Dependency.
  The following types/formats are supported:
    - Instances of Dependency.
    - String or CharSequence values, for example 'org.gradle:gradle-core:1.0'.
    - Maps, for example [group: 'org.gradle', name: 'gradle-core', version: '1.0'].
    - FileCollections, for example files('some.jar', 'someOther.jar').
    - Projects, for example project(':some:project:path').
    - ClassPathNotation, for example gradleApi().

Comment (tgroh) [reply runners/reference/java/build.gradle]: Fixed

**Pull request merged**